### PR TITLE
refactor: create conversation repository - WPB-8701

### PR DESCRIPTION
**Pull request by jullianm**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-8701" title="WPB-8701" target="_blank">TaskWPB-8701</a>  [iOS] Create ConversationRepository
  </td></table>
  <br />
 

### Key points

`ConversationRepository` acts as an in-between layer between the app’s business logic and data storage.
It contains methods to pull conversations, delete conversations, remove a self user from conversations and mark conversations as fetched.

Since the repository contains many CoreData related work (and got pretty massive), the PR introduces a new dedicated object called `ConversationLocalStore` that encapsulates all storage conversation related operations (using an injected context for now).
`ConversationLocalStore` conforms to the `ConversationLocalStoreProtocol` which makes the layer easily testable.

This enables us to isolate the storage layer in the same way we isolated the API layer (e.g `ConversationsAPI`) so that now the repository will not handle storage operations directly but rely on a dedicated local store object.

### Testing

Following use cases are covered:
- Pulling and storing conversations of different types (group, self, connection, 1:1)
- Deleting conversations
- Removing self user from conversations
- Marking conversations as fetched

### Checklist

- [ ] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [ ] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [ ] Make sure you use the API for UI elements that support large fonts.
- [ ] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [ ] New UI elements have Accessibility strings for VoiceOver.

### CODE REVIEW

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+/// An extension that encapsulates storage operations related to conversation group (user, role, members).
+
+extension ConversationLocalStore {
+
+    // MARK: - User & Role
+
+    func fetchUserAndRole(
+        from remoteConversationMember: WireAPI.Conversation.Member,
+        for localConversation: ZMConversation
```
Comment (johnxnguyen) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift:30]:
```suggestion
        for remoteConversationMember: WireAPI.Conversation.Member,
        in localConversation: ZMConversation
```

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+MLS.swift [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    ) {
+        guard let newMessageProtocol = remoteConversation.messageProtocol else {
+            return
+        }
+
+        localConversation.messageProtocol = newMessageProtocol.toDomainModel()
+    }
+
+    func updateMessageProtocol(
+        from remoteConversation: WireAPI.Conversation,
```
Comment (johnxnguyen) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+MLS.swift:43]: It looks like you can simply pass in the message protocol instead of the whole conversation.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationModelMappings.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireAPI
+import WireDataModel
+
+extension WireAPI.MLSCipherSuite {
+
+    func toDomainModel() -> WireDataModel.MLSCipherSuite {
+        .init(rawValue: rawValue)!
```
Comment (samwyndham) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationModelMappings.swift]: I see this can't fail in its current state but I'm wondering if it is worth adding safety here incase one of the enums changes. Consider adding a test for this that at a minimum asserts that all `WireAPI.MLSCipherSuite` cases can be converted to `WireDataModel.MLSCipherSuite`.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationModelMappings.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    }
+}
+
+extension WireAPI.ConversationAccessRoleLegacy {
+
+    func toDomainModel() -> WireDataModel.ConversationAccessRole? {
```
Comment (samwyndham) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationModelMappings.swift]: Does this need to be `optional`? Can you do:

```swift
    func toDomainModel() -> WireDataModel.ConversationAccessRole {
        switch self {
        case .private:
            .private
        case .team:
            .team
        case .activated:
            .activated
        case .nonActivated:
            .nonActivated
        }
    }
```

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+        self.context = context
+    }
+
+    // MARK: - Public
+
+    public func storeConversation(
+        _ conversation: WireAPI.Conversation,
+        isFederationEnabled: Bool
+    ) async throws {
+        guard let id = conversation.id ?? conversation.qualifiedID?.uuid else {
```
Comment (samwyndham) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift:98]: Is a missing ID a valid state? If no we should do something other than failing silently. Consider, throwing or logging the error and asserting. Also consider doing something in the default case below.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+        default:
+            /// conversation type is nil
+            return
+        }
+    }
+
+    public func deleteConversation(
+        withQualifiedId qualifiedId: WireAPI.QualifiedID
+    ) async throws {
+        let conversation = await context.perform { [context] in
```
Comment (samwyndham) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift]: Do we only care about group conversations? Maybe we should rename the method to show that?

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+
+    /// Stores the specified failed conversation
+    /// - Parameter qualifiedId: The conversation qualified ID.
+
+    func storeFailedConversation(
+        withQualifiedId qualifiedId: WireAPI.QualifiedID
+    ) async
+}
+
+public final class ConversationLocalStore: ConversationLocalStoreProtocol {
```
Comment (samwyndham) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift:60]: There are a lot of ways that these methods can return early. As someone looking at the code I'm not sure if this is expected behavior or if this just handling incorrect states. If it is correct behavior I think it should be documented, perhaps in the doc comments of the methods. If it is incorrect behavior, I think it should either throw or log the error & assert. Please revisit areas in this file and others and clarify expected behavior using throws, doc comments, logging etc

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+        guard let mlsService, let groupID else {
+            return
+        }
+
+        try await mlsService.wipeGroup(groupID)
+    }
+
+    public func removeSelfUserFromConversation(
+        withQualifiedId qualifiedId: WireAPI.QualifiedID
+    ) async {
```
Comment (samwyndham) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift]: I think this should throw if we attempt to delete a user that is not the self user.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    ) -> (ZMUser, Role?)? {
```
Comment (samwyndham) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift]: Consider adding parameter names to the tuple so that callers don't have to access results using `1` etc. - e.g. 

```swift
(user: ZMUser, role: Role?)
```

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+MLS.swift [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+            }
+        }
+    }
+
+    // MARK: - Self / MLS
+
+    func createOrJoinSelfConversation(
+        from localConversation: ZMConversation
+    ) async throws {
+        guard let context = localConversation.managedObjectContext else {
```
Comment (samwyndham) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+MLS.swift]: Can you not use the context that belongs to self here? That way it can't be `nil`.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift (hunk 6 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift]: I try to stick at best to what current legacy code is currently doing to avoid introducing breaking changes since these repo will soon be consumed (from my understanding) by existing code. So according to legacy code (`ConversationRequestStrategy` and `ConversationEventPayloadProcessor`), this only concerned `group` conversations however I can rename it `deleteConversationGroup` to be more explicit.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift (hunk 7 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift:60]: This code was moved from legacy class `ConversationEventPayloadProcessor ` to our new architecture. This code is, for the most part of it, not commented so I tried not to change too many things, just refactoring repetitive code, cleaning up code, making it more readable, more organized. 

The local store is just an object that encapsulates the already existing storage logic so right now it is supposed to behave like it was in legacy code. So, to answer your question, based on legacy code (given legacy code is correct), this is the expected behavior.

For instance, in legacy code we never handle any errors, just return early and proceed to the next conversation interation. 

```
    func updateOrCreateConverations(
        from payload: Payload.QualifiedConversationList,
        in context: NSManagedObjectContext
    ) async {
        for payload in payload.found {
            await updateOrCreateConversation(
                from: payload,
                source: .slowSync,
                in: context
            )
        }
    }
```

We do have loggers but from my understanding for now we removed these logs from our repositories.

 I will try to do some digging in order to be more explicit about each method's role.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+MLS.swift (hunk 10 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+MLS.swift]: will do thx

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift (hunk 9 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift]: yep 👍

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift (hunk 5 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift:98]: I will check that with backend team, thx

File: WireDomain/Tests/WireDomainTests/Helpers/XCTestCase+Assertions.swift [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    /// Assert that a collection is of a certain size.
+    ///
+    /// - Parameters:
+    ///   - collection: The collection to assert.
+    ///   - count: The expected number of elements.
+    ///   - message: The error message to show when the assertion fails.
+    ///   - file: The file name of the invoking test.
+    ///   - line: The line number when this assertion is made.
+
+    func XCTAssertCount(
```
Comment (johnxnguyen) [inline WireDomain/Tests/WireDomainTests/Helpers/XCTestCase+Assertions.swift]: Perhaps we can move this to WireTesting since it seems to be duplicated a few times.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationRepository.swift [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    func removeSelfUserFromConversations(
+        withQualifiedIds qualifiedIds: Set<WireAPI.QualifiedID>
+    ) async
+
+    /// Marks specified conversations as fetched.
+    /// - Parameter qualifiedIds: The qualified conversations IDs.
+
+    func markConversationsAsFetched(
+        qualifiedIds: Set<WireAPI.QualifiedID>
+    ) async
```
Comment (johnxnguyen) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationRepository.swift]: The ticket only specifies a need for pulling conversations. Is there a need for these right now?

Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationRepository.swift]: actually no, we're only dealing with slow sync so these are not needed at the moment, I will remove them (stash them for the future) and keep only the pullConversations method. 👍

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift (hunk 6 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift]: resolving this comment, as per John's comment, as we're dealing with slow sync in this PR, we don't need these methods for now, so I removed them

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift (hunk 8 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift]: resolving this comment, as per John's comment, as we're dealing with slow sync in this PR, we don't need these methods for now, so I removed this method, we'll deal with them later as the refactoring progress

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationRepository.swift (hunk 12 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationRepository.swift]: fixed, thanks, now we only have a pullConversations method, resolving this comment

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationModelMappings.swift (hunk 3 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationModelMappings.swift]: fixed

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationModelMappings.swift (hunk 4 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationModelMappings.swift]: fixed

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationModelMappings.swift (hunk 3 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationModelMappings.swift]: I added a full switch self, if we add a case, compilation will fail

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift (hunk 9 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift]: fixed

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+MLS.swift (hunk 10 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+MLS.swift]: fixed

File: WireDomain/Tests/WireDomainTests/Helpers/XCTestCase+Assertions.swift (hunk 11 above)
Comment (jullianm) [reply WireDomain/Tests/WireDomainTests/Helpers/XCTestCase+Assertions.swift]: fixed, these helpers were moved to WireTesting package

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift (hunk 5 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift:98]: Actually, we are using the WireAPI.Conversation object which is a new object that was recently created as part of the refactoring plan (see ticket https://wearezeta.atlassian.net/browse/WPB-8696 that is now closed) so we can assume that here we are consuming an object that is reliable.

Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift:98]: Based on legacy codebase, we're not doing anything here:

```
        guard let conversationType = payload.type.map(BackendConversationType.clientConversationType) else {
            return nil
        }
```

and 

```
        default:
            return nil
```

We just proceed to the next conversation iteration. 
I can add some print statement for now, please note that this ticket is just about slow sync and we will most definitely work again on these objects later on which will enable us to make some adjustments if needed.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift (hunk 7 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift:60]: I've added some documentation that hopefully will provide some general context on the different conversation types.

Comment (samwyndham) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift:60]: Thanks @jullianm !

**Review by samwyndham — APPROVED** [review]
🥳

File: WireAPI/Sources/WireAPI/Models/Conversation/ConversationType.swift [hunk 13]
```diff
@@ -20,7 +20,7 @@ import Foundation
 
 /// The various types of conversations.
 
-public enum ConversationType: Int, Codable {
+public enum ConversationType: Int, Codable, Sendable {
```
Comment (netbe) [inline WireAPI/Sources/WireAPI/Models/Conversation/ConversationType.swift:23]: is it for Swift 6?

File: WireDomain/Package.swift [hunk 14]
```diff
@@ -14,11 +14,19 @@ let package = Package(
     dependencies: [
         .package(url: "https://github.com/apple/swift-docc-plugin", from: "1.1.0"),
         .package(path: "../SourceryPlugin"),
-        .package(name: "WireAPI", path: "../WireAPI")
+        .package(name: "WireAPI", path: "../WireAPI"),
+        .package(name: "WireTestingPackage", path: "../WireTesting")
```
Comment (netbe) [inline WireDomain/Package.swift]: probably not related to this PR but could the name of the package be WireTesting ?

File: WireDomain/Package.swift [hunk 15]
```diff
@@ -14,11 +14,19 @@ let package = Package(
...
     ],
     targets: [
         .target(name: "WireDomainPkg", dependencies: ["WireAPI"], path: "./Sources/Package", swiftSettings: swiftSettings),
-        .testTarget(name: "WireDomainPkgTests", dependencies: ["WireDomainPkg"], path: "./Tests/PackageTests"),
+        .testTarget(
+            name: "WireDomainPkgTests",
```
Comment (netbe) [inline WireDomain/Package.swift]: probably not related to this PR but could the name of the package be WireDomainTests ?

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+            in: context
+        )
+
+        func fetchOrCreateRoleForConversation(
+            name: String,
+            conversation: ZMConversation
+        ) -> Role {
+            Role.fetchOrCreateRole(
+                with: name,
+                teamOrConversation: conversation.team != nil ? .team(conversation.team!) : .conversation(conversation),
```
Comment (netbe) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift]: there's a helper method in DataModel for this
```suggestion
                teamOrConversation: .fromTeamOrConversation(team: team, conversation: conversation),
```

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+        let selfUserRole = fetchUserAndRole(
+            from: members.selfMember,
+            for: localConversation
+        )?.role
+
+        localConversation.updateMembers(otherMembers, selfUserRole: selfUserRole)
+    }
+
+    // MARK: 1:1
```
Comment (netbe) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift]:
```suggestion
    // MARK: - 1:1
```

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Status.swift [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+
+    func updateConversationIfNeeded(
+        localConversation: ZMConversation,
+        fallbackGroupID: MLSGroupID?
+    ) async {
+        let (messageProtocol, mlsGroupID, mlsService) = await context.perform { [self] in
+            (
+                localConversation.messageProtocol,
+                localConversation.mlsGroupID,
+                context.mlsService
```
Comment (netbe) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Status.swift]: @jullianm did you discuss injecting mlsService directly with John? this is kind of technical debt we have that the mlsService is passed from NSManagedObjectContext.

But as we refactor this we can think of injecting it to at least the store directly instead

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift [hunk 19]
```diff
@@ -0,0 +1,437 @@
...
+    ///
+    /// __What is `Federation` ?__
+    ///
+    /// Businesses that use Wire in their organization have 2 choices, either deploy their teams in the Wire cloud or install an on-prem backend.
+    /// On-prem backends are isolated islands that before could not communicate with users from other on-prem islands or Wire cloud users.
+    /// Federation allows islands to be connected. Communication can be established between users from different backends the same way as on a single backend.
+    ///
+    /// - Parameter remoteConversation: The conversation object received from backend.
+    /// - Parameter removeConversationID: The conversation ID received from backend.
+    /// - Parameter isFederationEnabled: A flag indicating whether a federation is enabled.
```
Comment (netbe) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift]: we're repeating the definition of Federation, maybe we move this back to a docC within WireDomain and point to it instead

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationRepository.swift [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+            for conversation in foundConversations {
+                taskGroup.addTask { [self] in
+                    do {
+                        try await conversationsLocalStore.storeConversation(
+                            conversation,
+                            isFederationEnabled: backendInfo.isFederationEnabled
+                        )
+                    } catch {
+                        throw ConversationRepositoryError.failedToStoreConversation(error)
```
Comment (netbe) [inline WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationRepository.swift]: not sure here, does it break the loop at first error?

File: WireAPI/Sources/WireAPI/Models/Conversation/ConversationType.swift (hunk 13 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/Models/Conversation/ConversationType.swift:23]: yes, these are some the "easy" Swift 6 concurrency fixes : marking value types as Sendable.

File: WireDomain/Package.swift (hunk 14 above)
Comment (jullianm) [reply WireDomain/Package.swift]: No problem, will do so.

Comment (jullianm) [reply WireDomain/Package.swift]: We could do these specific changes in the PR but looking at the different Swift packages, I think we should streamline our Swift packages (WireUtilities, WireSystem, WireDomain,  WireTesting..) to the same naming convention. Making it more readable.

So, for instance: 

```
let package = Package(
    name: "WireDomainPackage",
    defaultLocalization: "en",
    platforms: [.iOS(.v15), .macOS(.v12)],
    products: [
        .library(name: "WireDomainPackage", type: .dynamic, targets: ["WireDomainPkg"]),
        .library(name: "WireDomainPackageSupport", type: .dynamic, targets: ["WireDomainPkgSupport"])
    ]

{...}
```

could become:

```
let package = Package(
    name: "WireDomain",
    defaultLocalization: "en",
    platforms: [.iOS(.v15), .macOS(.v12)],
    products: [
        .library(name: "WireDomain", type: .dynamic, targets: ["WireDomain"]),
        .library(name: "WireDomainSupport", type: .dynamic, targets: ["WireDomainSupport"])
    ]

{...}
```

And do the same for all packages.. I think it's better to have a dedicated ticket for this task as it concerns many Swift packages not related to this PR (I can work on this dedicated ticket)

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift (hunk 16 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Group.swift]: fixed, we can actually use the public static func `matching(_ conversation: ZMConversation) -> TeamOrConversation` but it does call the method you suggested. thanks.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Status.swift (hunk 18 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore+Status.swift]: no we didn't discuss it but it makes sense to inject it directly to the store, I fixed it thx

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationRepository.swift (hunk 20 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationRepository.swift]: Interestingly enough, it would do nothing because we are throwing inside a child task and not inside the body of `withThrowingTaskGroup` itself. If we were throwing inside the body, it would cancel the group and all of its child tasks but here it will do nothing and continue processing. I'm not a 100% sure but looking at Apple's doc and some online resources, it looks like it. Bottom line, if we want to `actually` throw and propagate the error to the call site, it should be done inside the body of `withThrowingTaskGroup` and not inside the child task.

Anyways, this was removed now that I added back the logger, we are catching the underlying error and logging it, so we don't need to have a throwing func anymore.

File: WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift (hunk 19 above)
Comment (jullianm) [reply WireDomain/Sources/WireDomain/Repositories/Conversations/ConversationLocalStore.swift]: done, I've added some documentation for conversations and federation using docc

File: WireDomain/Package.swift (hunk 14 above)
Comment (netbe) [reply WireDomain/Package.swift]: added a follow up ticket here https://wearezeta.atlassian.net/browse/WPB-10980

**Review by netbe — APPROVED** [review]
